ospec/ospec.js

```diff
@@ -143,9 +143,6 @@ else window.o = m()
 				}
 
 				if (fn.length > 0) {
-					var body = fn.toString()
-					var arg = (body.match(/\(([\w$]+)/) || body.match(/([\w$]+)\s*=>/) || []).pop()
```

Comment:
I'd rather fix the matching logic.

You could swap the regex order, and add a `^` to anchor what will then be the first one.

Reply:
I agree this should be shelved and redone 'properly'... In which circumstance I'd rather leave that particular change with your name in the git blame logs ;P - IMO this is very hacky stuff that even in the best case scenario steps beyond the kind of mechanical predictability one expects out of a simple test harness - it's on par with the magical ambition that had us trying to determine the notional 'end' of a promise chain for `m.request` redraw resolution at the start of the rewrite, suddenly straying unexpectedly far into the user's headspace. Like I said in #2158, I think this was a kind of compensation for the fact that async timeouts were never caught as part of normal reporting in the first place.

But now that I think of the thing more holistically, we could do well to fold the timing shortfall logs (`# elapsed: X, expected under Y`)  into the reports too.

Reply:
That totally makes sense, yes... With proper timeout messages this is indeed a lot less useful, but still nice to have (the more info we give users, the better).

Unlike the promise/redraw thing, parsing JS is well defined.

The `indexOf`/`lastIndexOf` thing is problematic though, because it can fail if `done` is present in a string or a comment, or even worse, a RegExp or a template literal (though the problem would then be caught by the timeout handler).

I say RegExps and template literals are worse because they can't be parsed with RegExps, unlike plain strings and comments.

Reply:
I'll fix the matching logic then no worries.

Reply:
The problem with this kind of cleverness is that it's maddening when it falls short of user intuition, and I just can't imagine a complete criteria that would rule out false positives & negatives.

The number of legitimate permutations of written code that can successfully resolve one of these is just bewildering. Suppose we supply the test with a stream or a bound array.push, on the proviso that the application mechanism under scrutiny will pull it out somewhere else and execute it pending whatever other conditions. The circumstances under which we can't say for sure that the argument is called somewhere is small, but the noise is akin to a particularly opinionated linter that isn't configurable.

More pertinent, the sense of safety this will cultivate in the useful cases sets an intuitive precedent that ospec can work out whether your code can feasibly run to completion, which is impossible to assure. It doesn't guard against logically unreachable code branches or mis-assigned references. I'm projecting a lot, but I can foresee a particularly pernicious kind of frustration when a user gets it in their head that this functionality offers some unprecedented insights into the tricky world of async, and after untold hours of hacking at presumably broken source under the conviction that ospec has validated the tests, eventually realises that the tests were never going to work in the first place.

The reason I mention the magical thinking 'conclusion of promise chain' is that the manner of ways in which async can be horribly confusing is colossal, and this really isn't helping with those fundamental problems, but thinks and suggests that it is. Far better IMO to introduce assurances like "you didn't specify any assertions for this test definition", which ospec is actually capable of asserting, rather than trying to sniff stringified user source in an attempt to make dubious assertions about the nature of callbacks.

Reply:
Eesh, I'm really ranting now and this sounds a bit nasty but I feel I have to add that the underlying criticism I'm making of this particular piece of code is that it was written at the same time as that chaining promise thing, at a time where all sorts of very neat but totally outside-remit API tricks were being used, I have to say, out of a naive exploration of the already contentious domain of "what's a good async API, and why might that be confusing". See also from around the time of rewrite: should we use promises at all; webpack does use promises and so does `onmatch`, etc etc.

Reply:
This is why we should've used an off-the-shelf testing framework and avoided all of the arguments and tech debt we're now saddled with by going it alone. 😑

Reply:
Yeah, and I feel a bit of a fool for my initial enthusiasm & satisfaction digging into this problem when I could have just been using tape. We've crossed the Rubicon of taking on external dependencies for a tool whose initial value proposition was that Mithril core could remain completely dependency free, and we're yak-shaving a library whose README talks about avoiding analysis paralysis.

Still think vocal conversation about maintenance gripes fears and hopes would be good to circumvent some of the alienation of slaving away through the GitHub process.

Reply:
I just realized there _could_ be false positives to the current approach, though you'd have to create them on purpose: `eval( "do" + "ne()")`...

I hadn't thought about the half-working detection imparting a false sense of security in programming newcomers. So let's get rid of it!

I suspect the mocks, the bundler and ospec were part of a devious ploy by @lhorie to publicly demonstrate the breadth of his abilities :-)

Reply:
@tivac thinking more about this, there's far more tech debt in the mocks and bundler (which are ad-hoc as can be, incomplete, quirky) than there is in ospec, which is IMO a fine piece of software (even if buggy at times, though v3 will be in a much better shape).